hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/ObjectInspectorCache.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.hadoop.hive.ql.io.parquet.serde.ArrayWritableObjectInspector;
+import org.apache.hadoop.hive.serde.serdeConstants;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+import org.apache.hadoop.io.ArrayWritable;
+import org.apache.hadoop.mapred.JobConf;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * To read value from an ArrayWritable, an ObjectInspector is needed.
+ * Object inspectors are cached here or created using the column type map.
+ */
+public class ObjectInspectorCache {
+  private final Map<String, TypeInfo> columnTypeMap = new HashMap<>();
+  private final Cache<Schema, ArrayWritableObjectInspector>
+      objectInspectorCache = Caffeine.newBuilder().maximumSize(1000).build();
+
+  public Map<String, TypeInfo> getColumnTypeMap() {
+    return columnTypeMap;
+  }
+
+  public ObjectInspectorCache(Schema tableSchema, JobConf jobConf) {
+    //From AbstractRealtimeRecordReader#prepareHiveAvroSerializer
+    // hive will append virtual columns at the end of column list. we should remove those columns.
+    // eg: current table is col1, col2, col3; jobConf.get(serdeConstants.LIST_COLUMNS): col1, col2, col3 ,BLOCK__OFFSET__INSIDE__FILE ...
+    Set<String> writerSchemaColNames = tableSchema.getFields().stream().map(f -> f.name().toLowerCase(Locale.ROOT)).collect(Collectors.toSet());
+    List<String> columnNameList = Arrays.stream(jobConf.get(serdeConstants.LIST_COLUMNS).split(",")).collect(Collectors.toList());
+    List<TypeInfo> columnTypeList =  TypeInfoUtils.getTypeInfosFromTypeString(jobConf.get(serdeConstants.LIST_COLUMN_TYPES));
+
+    int columnNameListLen = columnNameList.size() - 1;
+    for (int i = columnNameListLen; i >= 0; i--) {
+      String lastColName = columnNameList.get(columnNameList.size() - 1);
+      // virtual columns will only append at the end of column list. it will be ok to break the loop.
+      if (writerSchemaColNames.contains(lastColName)) {
+        break;
+      }
+      columnNameList.remove(columnNameList.size() - 1);
+      columnTypeList.remove(columnTypeList.size() - 1);
+    }
+
+    //Use columnNameList.size() instead of columnTypeList because the type list is longer for some reason
+    IntStream.range(0, columnNameList.size()).boxed().forEach(i -> columnTypeMap.put(columnNameList.get(i),
+        TypeInfoUtils.getTypeInfosFromTypeString(columnTypeList.get(i).getQualifiedName()).get(0)));
+
+    StructTypeInfo rowTypeInfo = (StructTypeInfo) TypeInfoFactory.getStructTypeInfo(columnNameList, columnTypeList);
+    ArrayWritableObjectInspector objectInspector = new ArrayWritableObjectInspector(rowTypeInfo);
```

Comment:
@xicm @xiarixiaoyao can you help confirm the correctness?

Reply:
There may be compatibility issues between hive2 and hive3. DATE, TIMESTAMP

Reply:
> There may be compatibility issues between hive2 and hive3. DATE, TIMESTAMP

I think hive will handle this itself.

Reply:
FYI this is pretty much a copy of https://github.com/apache/hudi/blob/e9389ffde53fa2b28feba248b7e8f17fd565e458/hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java#L111

Reply:
I test with hive3, this works well.